From 277e4a2308f6690f0a2c763aa47da1b17bc51a43 Mon Sep 17 00:00:00 2001
From: James Haggerty <james.haggerty@morsemicro.com>
Date: Thu, 10 Apr 2025 16:35:38 +1000
Subject: [PATCH 1/2] APP-4377 Change RD09 RX LED to be a TPT LED instead

This is the default configuration for most wifi chips (e.g. iwlwifi/mt76/ath9k),
and avoids the unfortunate regression in the 5.13 kernel which causes
the RX LED to flash on every beacon it sees.

morse_led_init has to be moved since you have to register the tpt
LED before you do ieee80211_register_hw. Apparently, if you don't
do this as the documentation specifies, you get a kernel crash
(how do I know this, since surely I was following the doco?).

Please discuss appropriateness of thresholds.
---
 led.c | 23 ++++++++++++++++++++++-
 led.h |  2 ++
 mac.c |  4 ++--
 3 files changed, 26 insertions(+), 3 deletions(-)

diff --git a/led.c b/led.c
index 2ee0c9a1..f21a6ccc 100644
--- a/led.c
+++ b/led.c
@@ -11,6 +11,20 @@
 #include <linux/leds.h>
 #include <linux/workqueue.h>
 
+/* This is based on the iwlwifi.c blink rates. */
+static const struct ieee80211_tpt_blink morse_blink[] = {
+	{ .throughput = 0, .blink_time = 334 },
+	{ .throughput = 512 - 1, .blink_time = 260 },
+	{ .throughput = 1024 - 1, .blink_time = 220 },
+	{ .throughput = 2 * 1024 - 1, .blink_time = 190 },
+	{ .throughput = 3 * 1024 - 1, .blink_time = 170 },
+	{ .throughput = 6 * 1024 - 1, .blink_time = 150 },
+	{ .throughput = 12 * 1024 - 1, .blink_time = 130 },
+	{ .throughput = 18 * 1024 - 1, .blink_time = 110 },
+	{ .throughput = 24 * 1024 - 1, .blink_time = 80 },
+	{ .throughput = 30 * 1024 - 1, .blink_time = 50 },
+};
+
 static void morse_led_callback(struct led_classdev *led_cdev, enum led_brightness brightness)
 {
 	struct morse_led *led = container_of(led_cdev, struct morse_led, led_cdev);
@@ -78,6 +92,12 @@ int morse_led_register(struct morse *mors, struct morse_led *led, int pin_num,
 	case MORSE_LED_RADIO:
 		led->led_cdev.default_trigger = ieee80211_get_radio_led_name(mors->hw);
 		break;
+	case MORSE_LED_TPT:
+		led->led_cdev.default_trigger =
+			ieee80211_create_tpt_led_trigger(mors->hw,
+					IEEE80211_TPT_LEDTRIG_FL_CONNECTED,
+					morse_blink, ARRAY_SIZE(morse_blink));
+		break;
 	case MORSE_LED_USER_DEF:
 		/* Still register led but defer control to user space */
 		led->led_cdev.default_trigger = NULL;
@@ -122,7 +142,7 @@ void morse_led_init(struct morse *mors)
 	led_group->num_leds = NUM_LEDS;
 
 	morse_led_register(mors, &led_group->leds[0], LED_G, "morse_led0", MORSE_LED_ASSOC);
-	morse_led_register(mors, &led_group->leds[1], LED_R, "morse_led1", MORSE_LED_RX);
+	morse_led_register(mors, &led_group->leds[1], LED_R, "morse_led1", MORSE_LED_TPT);
 	morse_led_register(mors, &led_group->leds[2], LED_PWR, "morse_led2", MORSE_LED_HOST);
 }
 
@@ -146,6 +166,7 @@ void morse_led_exit(struct morse *mors)
 		case MORSE_LED_RX:
 		case MORSE_LED_ASSOC:
 		case MORSE_LED_RADIO:
+		case MORSE_LED_TPT:
 		case MORSE_LED_USER_DEF:
 			led_classdev_unregister(&led->led_cdev);
 			cancel_work_sync(&led->work);
diff --git a/led.h b/led.h
index a18f059f..d4422356 100644
--- a/led.h
+++ b/led.h
@@ -24,6 +24,7 @@ struct morse;
  *		correctly initialised
  * The remaining five options register the LED with the kernel subsystem LED callbacks
  *	- TX and RX indicate there is network traffic
+ *	- TPT flashes based on traffic with faster flashing for higher throughput
  *	- ASSOC indicates that the station has associated (STA only)
  *	- RADIO indicates that the RF is up/down
  *	- USER_DEF registers the LED with mac80211 but does not set a default behaviour
@@ -33,6 +34,7 @@ struct morse;
 enum led_mode {
 	MORSE_LED_DISABLED,
 	MORSE_LED_HOST,
+	MORSE_LED_TPT,
 	MORSE_LED_TX,
 	MORSE_LED_RX,
 	MORSE_LED_ASSOC,
diff --git a/mac.c b/mac.c
index 828897df..cef1ff71 100644
--- a/mac.c
+++ b/mac.c
@@ -6940,6 +6940,8 @@ int morse_mac_register(struct morse *mors)
 		goto err;
 	}
 
+	morse_led_init(mors);
+
 	/* We manage our own regdb, as Linux has no S1G support yet */
 	mors->wiphy->regulatory_flags = REGULATORY_WIPHY_SELF_MANAGED;
 	mors->wiphy->reg_notifier = morse_reg_notifier;
@@ -7017,8 +7019,6 @@ int morse_mac_register(struct morse *mors)
 		morse_log_modparams(mors);
 #endif
 
-	morse_led_init(mors);
-
 	return ret;
 
 #ifdef CONFIG_MORSE_RC
-- 
2.43.0

